### PR TITLE
VarMap

### PR DESCRIPTION
**Pull request by dcousens**
### VarMap(lengthCodec, keyCodec, valueCodec)

``` javascript
vstruct([
   ['items', vstruct.VarMap(vstruct.UInt16, vstruct.VarString(vstruct.UInt8), vstruct.UInt32LE)]
])
```

You could then pass an arbitrary JS object (mapping to `vstruct.UInt32LE` values, or some other codec) and have an optimal encoding format for that object.

I'm thinking it would just turn into equivalent of `vsturct.VarArray`,  but it would unpack the `key`/`value` pairs for you.

### CODE REVIEW

**Comment by dcousens** [comment]
Had to fix a bug in `string` where it would encode a string longer than allowed... see https://github.com/varstruct/varstruct/pull/28/commits/4e1edd8e5faab27f178fd59731f5cbaa646d1c73

**Comment by dcousens** [comment]
added d5307d9 to pass the tests (due to coverage issues?)

**Comment by dcousens** [comment]
@fanatid I'm not sure why but Node version 4 is blowing up... https://travis-ci.org/varstruct/varstruct/jobs/226719224

``` javascript
Buffer.from('deadbeef', 'hex').compare(Buffer.from('deadbeefffff', 'hex'), 0, 4)
```

**Comment by dcousens** [comment]
tests are only failing now due to the issue in `Value`... good to merge,  opened https://github.com/varstruct/varstruct/issues/29 for the existing issue

File: lib/varmap.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  if (!util.isAbstractCodec(lengthType)) throw new TypeError('lengthType is invalid codec')
+  if (!util.isAbstractCodec(keyType)) throw new TypeError('keyType is invalid codec')
+  if (!util.isAbstractCodec(valueType)) throw new TypeError('valueType is invalid codec')
+
+  function _length (object) {
+    if (!object) throw new TypeError('Expected object')
+
+    var size = 0
+    var i = 0
+    for (var key in object) {
```
Comment (fanatid) [inline lib/varmap.js]: Why not `Object.keys`?

**Comment by fanatid** [comment]
please rebase
other lgtm (except `for-in` instead `Object.keys`, why?)

File: lib/varmap.js (hunk 1 above)
Comment (dcousens) [inline lib/varmap.js]: cause I needed both the key and the object... why `Object.keys`?

Comment (fanatid) [inline lib/varmap.js]: I meant, why not `for (var i = 0, keys = Object.keys(object); i < keys.length; ++i)`

Comment (dcousens) [inline lib/varmap.js]: whats the advantage though?
It just seems like more code/syntax?

Comment (fanatid) [inline lib/varmap.js]: they have difference:
```js
function A () { this.a = 1 }
A.prototype.b = 2
const a = new A()

console.log(a) // A { a: 1 }
console.log(Object.keys(a)) // [ 'a' ]
for (const x in a) console.log(x) // a, b
```

Comment (dcousens) [inline lib/varmap.js]: Right,  is that an issue?

Comment (fanatid) [inline lib/varmap.js]: no, I just curious why you used `for-in` here

Comment (dcousens) [inline lib/varmap.js]: @fanatid I suppose I figure I've never personally found a situation where I pass an object that has a `prototype` to this sort of function;  but I also avoid complex classes wherever possible.

Comment (fanatid) [inline lib/varmap.js]: if you okay with current `for-in` then let's stay with this

**Comment by dcousens** [comment]
Merged via rebase

**Pull request merged**